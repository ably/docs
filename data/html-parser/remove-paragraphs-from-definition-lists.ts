export const removeParagraphsFromDefinitionListsAndMerge = (cheerioNodes: cheerio.Selector) =>
  cheerioNodes('dl').each((_i: number, elem: cheerio.Element) => {
    // As jQuery/Cheerio is stateful and imperative, including comments to clarify
    // Remove empty paragraphs immediately after any dl
    const siblings = cheerioNodes(elem)
      .siblings('p')
      .filter((_, elem) => {
        const self = cheerioNodes(elem);
        return self.text().trim() === '';
      });
    if (siblings.length) {
      siblings.remove();
    }

    // Now merge multiple consecutive dls
    const nextDl = cheerioNodes(elem).nextUntil(':not(dl)');
    cheerioNodes(elem).append(nextDl.children());
    nextDl.remove();
  });
