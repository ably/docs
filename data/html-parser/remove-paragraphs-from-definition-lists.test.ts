import cheerio from 'cheerio';
import { removeParagraphsFromDefinitionListsAndMerge } from './remove-paragraphs-from-definition-lists';

const consecutiveDls = `<dl>
  <dt>Term One</dt>
  <dd>Definition One</dd>
</dl>
<dl>
  <dt>Term Two</dt>
  <dd>Definition Two</dd>
  <dt>Term Two</dt>
  <dd>Definition Three</dd>
</dl>`;

const consecutiveDlsWithPElement = `<dl>
  <dt>Term One</dt>
  <dd>Definition One</dd>
</dl>
<p></p>
<dl>
  <dt>Term Two</dt>
  <dd>Definition Two</dd>
  <dt>Term Two</dt>
  <dd>Definition Three</dd>
</dl>`;

describe('Consecutive Dl elements are merged', () => {
  it('Merges consecutive Dl elements', () => {
    const loadedDom = cheerio.load(consecutiveDls);
    removeParagraphsFromDefinitionListsAndMerge(loadedDom);
    expect(loadedDom.html()).toBe(`<html><head></head><body><dl>
  <dt>Term One</dt>
  <dd>Definition One</dd>
<dt>Term Two</dt><dd>Definition Two</dd><dt>Term Two</dt><dd>Definition Three</dd></dl>
</body></html>`);
  });
  it('Merges consecutive Dl elements with intervening empty p element', () => {
    const loadedDom = cheerio.load(consecutiveDlsWithPElement);
    removeParagraphsFromDefinitionListsAndMerge(loadedDom);
    expect(loadedDom.html()).toBe(`<html><head></head><body><dl>
  <dt>Term One</dt>
  <dd>Definition One</dd>
<dt>Term Two</dt><dd>Definition Two</dd><dt>Term Two</dt><dd>Definition Three</dd></dl>

</body></html>`);
  });
});
