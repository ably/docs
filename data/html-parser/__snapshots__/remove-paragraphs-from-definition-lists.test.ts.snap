// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Consecutive Dl elements are merged Renders an image if a dl is on the same page 1`] = `
LoadedCheerio {
  "0": <dl>
    
    	
    <dt>
      Occupancy
    </dt>
    
    	
    <dd>
      provides metrics about the clients attached to a channel. This includes a count of the number of clients attached, the number of publishers and subscribers, the number of presence members, and the number of presence publishers and subscribers.
    </dd>
    
    	
    <dt>
      Presence
    </dt>
    
    	
    <dd>
      provides information about the activity of members entered into the presence set of a channel. Presence members announce their state to presence subscribers via presence events. For example, when entering the presence set, leaving the presence set, or updating their member data.
    </dd>
    
    
  </dl>,
  "_root": LoadedCheerio {
    "0": Document {
      "children": Array [
        <html>
          <head />
          <body>
            <p>
              Ably’s presence feature allows clients or devices to announce their presence on a channel. Other devices or services may then subscribe to these presence events (such as entering, updating their state, or leaving the channel) in real time using our 
              <a
                href="\\\\\\"/realtime\\\\\\""
              >
                realtime SDKs
              </a>
              , or via the 
              <a
                href="\\\\\\"/general/integrations\\\\\\""
              >
                Ably Integrations
              </a>
              . You can also request a list of clients or devices that are online/offline on a channel at a particular point in time via the 
              <a
                href="\\\\\\"/api/rest-api#presence\\\\\\""
              >
                <span
                  class="\\\\\\"caps\\\\\\""
                >
                  REST
                </span>
                 
                <span
                  class="\\\\\\"caps\\\\\\""
                >
                  API
                </span>
              </a>
              .
            </p>
            
    
            <p>
              <a
                href="\\\\\\"/images/diagrams/Channels-Presence.gif\\\\\\""
                target="\\\\\\"_blank\\\\\\""
              >
                
     
                <img
                  alt="\\\\\\"Presence"
                  representation\\"=""
                  src="\\\\\\"/images/diagrams/Channels-Presence.gif\\\\\\""
                />
              </a>
            </p>
            
    
            <p>
              Furthermore, if persistence is enabled on the presence channel, you can also retrieve 
              <a
                href="\\\\\\"/api/rest-sdk/history#presence-history\\\\\\""
              >
                presence history
              </a>
               for the channel, i.e, static data about historical presence states of your clients/devices. This operation also can be done using both Ably’s 
              <a
                href="\\\\\\"/realtime\\\\\\""
              >
                Realtime
              </a>
               and 
              <a
                href="\\\\\\"/rest\\\\\\""
              >
                <span
                  class="\\\\\\"caps\\\\\\""
                >
                  REST
                </span>
              </a>
               libraries.
            </p>
            
    
            <h2
              id="\\\\\\"occupancy-presence\\\\\\""
            >
              Occupancy and presence
            </h2>
            
    
            <p>
              The occupancy and presence features differ in the following ways:
            </p>
            
    
            <dl>
              
    	
              <dt>
                Occupancy
              </dt>
              
    	
              <dd>
                provides metrics about the clients attached to a channel. This includes a count of the number of clients attached, the number of publishers and subscribers, the number of presence members, and the number of presence publishers and subscribers.
              </dd>
              
    	
              <dt>
                Presence
              </dt>
              
    	
              <dd>
                provides information about the activity of members entered into the presence set of a channel. Presence members announce their state to presence subscribers via presence events. For example, when entering the presence set, leaving the presence set, or updating their member data.
              </dd>
              
    
            </dl>
          </body>
        </html>,
      ],
      "endIndex": null,
      "next": null,
      "parent": null,
      "prev": null,
      "startIndex": null,
      "type": "root",
      "x-mode": "quirks",
    },
    "_root": [Circular],
    "length": 1,
    "options": Object {
      "decodeEntities": true,
      "xml": false,
    },
  },
  "length": 1,
  "options": Object {
    "decodeEntities": true,
    "xml": false,
  },
  "prevObject": LoadedCheerio {
    "0": Document {
      "children": Array [
        <html>
          <head />
          <body>
            <p>
              Ably’s presence feature allows clients or devices to announce their presence on a channel. Other devices or services may then subscribe to these presence events (such as entering, updating their state, or leaving the channel) in real time using our 
              <a
                href="\\\\\\"/realtime\\\\\\""
              >
                realtime SDKs
              </a>
              , or via the 
              <a
                href="\\\\\\"/general/integrations\\\\\\""
              >
                Ably Integrations
              </a>
              . You can also request a list of clients or devices that are online/offline on a channel at a particular point in time via the 
              <a
                href="\\\\\\"/api/rest-api#presence\\\\\\""
              >
                <span
                  class="\\\\\\"caps\\\\\\""
                >
                  REST
                </span>
                 
                <span
                  class="\\\\\\"caps\\\\\\""
                >
                  API
                </span>
              </a>
              .
            </p>
            
    
            <p>
              <a
                href="\\\\\\"/images/diagrams/Channels-Presence.gif\\\\\\""
                target="\\\\\\"_blank\\\\\\""
              >
                
     
                <img
                  alt="\\\\\\"Presence"
                  representation\\"=""
                  src="\\\\\\"/images/diagrams/Channels-Presence.gif\\\\\\""
                />
              </a>
            </p>
            
    
            <p>
              Furthermore, if persistence is enabled on the presence channel, you can also retrieve 
              <a
                href="\\\\\\"/api/rest-sdk/history#presence-history\\\\\\""
              >
                presence history
              </a>
               for the channel, i.e, static data about historical presence states of your clients/devices. This operation also can be done using both Ably’s 
              <a
                href="\\\\\\"/realtime\\\\\\""
              >
                Realtime
              </a>
               and 
              <a
                href="\\\\\\"/rest\\\\\\""
              >
                <span
                  class="\\\\\\"caps\\\\\\""
                >
                  REST
                </span>
              </a>
               libraries.
            </p>
            
    
            <h2
              id="\\\\\\"occupancy-presence\\\\\\""
            >
              Occupancy and presence
            </h2>
            
    
            <p>
              The occupancy and presence features differ in the following ways:
            </p>
            
    
            <dl>
              
    	
              <dt>
                Occupancy
              </dt>
              
    	
              <dd>
                provides metrics about the clients attached to a channel. This includes a count of the number of clients attached, the number of publishers and subscribers, the number of presence members, and the number of presence publishers and subscribers.
              </dd>
              
    	
              <dt>
                Presence
              </dt>
              
    	
              <dd>
                provides information about the activity of members entered into the presence set of a channel. Presence members announce their state to presence subscribers via presence events. For example, when entering the presence set, leaving the presence set, or updating their member data.
              </dd>
              
    
            </dl>
          </body>
        </html>,
      ],
      "endIndex": null,
      "next": null,
      "parent": null,
      "prev": null,
      "startIndex": null,
      "type": "root",
      "x-mode": "quirks",
    },
    "_root": [Circular],
    "length": 1,
    "options": Object {
      "decodeEntities": true,
      "xml": false,
    },
  },
}
`;
